tcases-lib/src/main/java/org/cornutum/tcases/io/SystemInputDocWriter.java

```diff
@@ -0,0 +1,373 @@
+//////////////////////////////////////////////////////////////////////////////
+// 
+//                    Copyright 2018, Cornutum Project
+//                             www.cornutum.org
+//
+//////////////////////////////////////////////////////////////////////////////
+
+package org.cornutum.tcases.io;
+
+import org.cornutum.tcases.*;
+import org.cornutum.tcases.conditions.*;
+import org.cornutum.tcases.util.XmlWriter;
+import static org.cornutum.tcases.VarValueDef.Type.*;
+import static org.cornutum.tcases.io.SystemInputDoc.*;
+import static org.cornutum.tcases.util.CollectionUtils.toStream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Writer;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Writes a {@link SystemInputDef} in the form of an XML document.
+ *
+ */
+public class SystemInputDocWriter extends AbstractSystemInputWriter
+  {
+  /**
+   * Creates a new SystemInputDocWriter object that writes to standard output.
+   */
+  public SystemInputDocWriter()
+    {
+    super();
+    }
+  
+  /**
+   * Creates a new SystemInputDocWriter object that writes to the given stream.
+   */
+  public SystemInputDocWriter( OutputStream stream)
+    {
+    super( stream);
+    }
+  
+  /**
+   * Creates a new SystemInputDocWriter object that writes to the given stream.
+   */
+  public SystemInputDocWriter( Writer writer)
+    {
+    super( writer);
+    }
+
+  /**
+   * Writes the given system test definition the form of an XML document.
+   */
+  public void write( SystemInputDef systemInput)
+    {
+    xmlWriter_.writeDeclaration();
+
+    xmlWriter_.writeTagStart( SYSTEM_TAG);
+    xmlWriter_.writeAttribute( NAME_ATR, systemInput.getName());
+    xmlWriter_.writeTagEnd();
```

Comment:
you could consider refactor such sequences into a single writeTag method for readability in all Writers.

Reply:
Something like

     public void writeTag(String tag, Map<String, String> attributes, Runnable contentSupplier)
      {
      writeTagStart(tag);
      if (attributes != null) 
      {
        attributes.forEach( this::writeAttribute);
      }
      if (contentSupplier != null) 
      {
        writeTagEnd();
        indent();
        contentSupplier.run();
        unindent();
        writeElementEnd(tag);
      } else 
        {
        writeEmptyElementEnd();
      }
    }

It's not perfect, but it can dramatically reduce the code of the writers and let you stop worrying about the correct place to end tags, elements, indents.

Reply:
when combined with Collections.singletonMap()  or some copy of Java9's Maps.of(...) o the caller side.

Reply:
Also using lambdas on the caller side for the body. 

     xmlWriter_.writeTag(SYSTEM_TAG, mapOf(NAME_ATR, systemInput.getName()), () -> {
        // produce tag body
      writeAnnotations( systemInput);
      toStream( systemInput.getFunctionInputDefs()).forEach( function -> writeFunction( function));
     });

Reply:
That way you could also make most XwmlWriter methods private and thus reduce coupling between classes

Reply:
These are all good suggestions. This code originated at a time before Java 8 and lambda expressions. It's probably time for an update and a cleanup.